bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,440 @@
+     * range.
+     */
+    private static class Reader extends AbstractBoundedReader<Result> {
+      private final BoundedSource<Result> source;
+      private final Scan scan;
+      private final CloudBigtableScanConfiguration config;
+
+      private Connection connection;
+      private ResultScanner scanner;
+      private Table table;
+
+      private Result current;
+
+      public Reader(BoundedSource<Result> source, CloudBigtableScanConfiguration config,
+          byte[] startRow, byte[] endRow) throws IOException {
+        super();
+        this.source = source;
+        this.config = config;
+        if (startRow != null || endRow != null) {
+          this.scan = new Scan(config.getScan());
+          this.scan.setStartRow(startRow);
+          this.scan.setStopRow(endRow);
+        } else {
+          this.scan = config.getScan();
+        }
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        current = scanner.next();
+        return current != null;
+      }
+
+      @Override
+      public void close() throws IOException {
+        scanner.close();
+        table.close();
+        connection.close();
+      }
+
+      @Override
+      public Result getCurrent() throws NoSuchElementException {
+        return current;
+      }
+
+      @Override
+      public BoundedSource<Result> getCurrentSource() {
+        return source;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        connection = new BigtableConnection(config.toHBaseConfig());
+        table = connection.getTable(TableName.valueOf(config.getTable()));
+        scanner = table.getScanner(scan);
+        return advance();
+      }
+    }
+
+    public Source(CloudBigtableScanConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public Coder<Result> getDefaultOutputCoder() {
+      return new HBaseResultCoder();
+    }
+
+    @Override
+    public List<BoundedSource<Result>> splitIntoBundles(
+        long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+      TableName tableName = TableName.valueOf(configuration.getTable());
+      try (Connection conn = new BigtableConnection(configuration.toHBaseConfig());
+          Admin admin = conn.getAdmin(); RegionLocator locator = conn.getRegionLocator(tableName)) {
+        List<BoundedSource<Result>> splits = new ArrayList<>();
+        Pair<byte[][], byte[][]> startEndKeys = locator.getStartEndKeys();
+
+        byte[][] startKeys = startEndKeys.getFirst();
+        byte[][] endKeys = startEndKeys.getSecond();
+        for (int i = 0; i < startKeys.length; i++) {
+          splits.add(new SourceWithKeys(configuration, startKeys[i], endKeys[i]));
+        }
+        return splits;
+      }
+    }
+
+    @Override
+    public void validate() {
+      checkNotNullOrEmpty(configuration.getProjectId(), "project");
```

Comment:
Can you validate if the project, zone, cluster, and table are actually valid here ?